format/spec.md

```diff
@@ -241,7 +245,14 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for field ids which are not present in a data file must be resolved according the following rules:
```

Comment:
Above says "All columns must be written to data file".
Here it says "field ids which are not present in a data file ..."
Besides schema evolution (new columns) case,  and multiple partitionings in use, when can the field ids  be missing from the data file?

Reply:
When files are added from Hive tables or other places. If you're writing to an Iceberg table we have stricter standards than if you are adding data files from a converted table.

In the past, we didn't document the expectations for readers, but this is definitely a case where we want to be clear about the requirements across engines.

Reply:
When migrating from other table formats that don't write this information into data files.  Hive is the primary example.

Reply:
I get it -- files migrated from hive lack some information and we need to deal with it.
If such "partial data files" are legal, why do we expect writers not to create them? The readers still need to support them, so what's the win (from format perspective)?

Reply:
The argument here is that especially with default values there is a larger room for misinterpretation.  I don't have a strong feeling here.

Reply:
Files are the source of truth for row data. It's convenient to have a copy of identity columns in metadata because we can avoid projecting and doing extra work. But you should still get a full copy of the row as it was written if you read just the file.